tls/s2n_signature_algorithms.c

```diff
@@ -226,81 +156,170 @@ S2N_RESULT s2n_signature_algorithm_recv(struct s2n_connection *conn, struct s2n_
+}
+
+static bool s2n_signature_algorithm_is_supported_by_peer(
+        struct s2n_connection *conn, uint16_t iana)
+{
+    return s2n_result_is_ok(s2n_signature_algorithms_validate_supported_by_peer(conn, iana));
 }
 
-int s2n_choose_sig_scheme_from_peer_preference_list(struct s2n_connection *conn,
-        struct s2n_sig_scheme_list *peer_wire_prefs,
-        const struct s2n_signature_scheme **sig_scheme_out)
+S2N_RESULT s2n_signature_algorithm_select(struct s2n_connection *conn)
 {
-    POSIX_ENSURE_REF(conn);
-    POSIX_ENSURE_REF(sig_scheme_out);
+    RESULT_ENSURE_REF(conn);
+    RESULT_ENSURE_REF(conn->secure);
+    struct s2n_cipher_suite *cipher_suite = conn->secure->cipher_suite;
+    RESULT_ENSURE_REF(cipher_suite);
 
-    const struct s2n_signature_scheme *chosen_scheme = &s2n_null_sig_scheme;
-    if (conn->actual_protocol_version < S2N_TLS13) {
-        POSIX_GUARD(s2n_choose_default_sig_scheme(conn, &chosen_scheme, conn->mode));
+    const struct s2n_signature_scheme **chosen_sig_scheme = NULL;
+    if (conn->mode == S2N_CLIENT) {
+        chosen_sig_scheme = &conn->handshake_params.client_cert_sig_scheme;
     } else {
-        /* Pick a default signature algorithm in TLS 1.3 https://tools.ietf.org/html/rfc8446#section-4.4.2.2 */
-        POSIX_GUARD(s2n_tls13_default_sig_scheme(conn, &chosen_scheme));
+        chosen_sig_scheme = &conn->handshake_params.server_cert_sig_scheme;
     }
 
-    /* SignatureScheme preference list was first added in TLS 1.2. It will be empty in older TLS versions. */
-    if (conn->actual_protocol_version >= S2N_TLS12 && peer_wire_prefs != NULL && peer_wire_prefs->len > 0) {
-        /* Use a best effort approach to selecting a signature scheme matching client's preferences */
-        POSIX_GUARD(s2n_choose_sig_scheme(conn, peer_wire_prefs, &chosen_scheme));
+    /* Before TLS1.2, signature algorithms were fixed instead of negotiated */
+    if (conn->actual_protocol_version < S2N_TLS12) {
+        RESULT_GUARD(s2n_signature_algorithms_get_legacy_default(conn, conn->mode, chosen_sig_scheme));
+        return S2N_RESULT_OK;
     }
 
-    *sig_scheme_out = chosen_scheme;
-    return S2N_SUCCESS;
+    const struct s2n_signature_preferences *signature_preferences = NULL;
+    RESULT_GUARD_POSIX(s2n_connection_get_signature_preferences(conn, &signature_preferences));
+    RESULT_ENSURE_REF(signature_preferences);
+
+    const struct s2n_signature_scheme *fallback_candidate = NULL;
+
+    /* We use local preference order, not peer preference order, so we iterate
+     * over the local preferences instead of over the options offered by the peer.
+     */
+    for (size_t i = 0; i < signature_preferences->count; i++) {
+        const struct s2n_signature_scheme *candidate = signature_preferences->signature_schemes[i];
+
+        if (!s2n_signature_scheme_is_valid_for_recv(conn, candidate)) {
+            continue;
+        }
+
+        if (s2n_is_sig_scheme_valid_for_auth(conn, candidate) != S2N_SUCCESS) {
```

Comment:
I'm assuming this check or the one above enforces that this signature is only used in the correct TLS version? There's quite a few comments below about TLS versions so I was expecting a check above and didn't find it.

Reply:
The one above does-- it checks if the scheme is valid to choose, which includes matching the current version.

Reply:
Can you add a comment for that on that check, just for future reviews 🙂